시그널링 서버 disconnect 로직 및 client에서 동작가능하도록 하는 로직을 추가한다

inspirit941 opened the pull request:
> resolve #2 

## Detail & Solution
socket 서버에서 peer disconnection 이벤트 발생시 프론트에서 처리되도록 해야합니다.

## What I did
disconnect 시의 로직 생성. 아직 프론트엔드에 disconnect 요청로직이 없어서 확인은 못했습니다.
해당 브랜치로 pull 받아서 테스트해보고 리뷰 부탁드립니다.

## What I need to do

inspirit941 committed:
feat: 🎸 disconnect login

disconnect 시의 로직 생성

inspirit941 committed:
fix: 🐛 cors 에러수정, 오탈자처리

dididy commented:
동건님과 CORS 이슈 늪에서 허우적 거리는 중..

voidmatcha committed:
Revert "fix: 🐛 cors 에러수정, 오탈자처리"

This reverts commit ae03051ae6c3a0f0428981555d5b758bc7952492.

voidmatcha committed:
refactor: 💡 node socket 시그널링 서버 동작 정상화 완료

- 아직 disconnect 부분 좀 더 정상 작동하도록 해야 함

voidmatcha committed:
refactor: 💡 클라이언트 WebRTC peer stream 연결 정상화

- 동건님이 수정하신 코드 기반으로 동작 가능하도록 수정
- disconnect 이벤트에서 문제가 생긴 부분 확인해봐야함

voidmatcha committed:
style: 💄 mock-server에 필요없는 코드 삭제

- http-proxy-middleware 삭제
- const { userInfo } = require('os'); 삭제

voidmatcha committed:
refactor: 💡 node socket 시그널링 서버 origin및 헤더 설정 추가

- 갑자기 오류가 떠서 해당 옵션을 추가해줘야 동작함

voidmatcha committed:
feat: 🎸 소켓 시그널링 방 찼을 때 첫화면으로 돌아가도록 처리

- 서버에서 emit된걸 클라이언트에서 처리하는 코드가 없었음

voidmatcha committed:
fix: 🐛 소켓 시그널링 서버 client disconnect시 users 객체 처리 추가

- 이부분 처리하지 않으면 users의 roomId에 user가 머물러있음

voidmatcha committed:
fix: 🐛 클라이언트 사용되지 않는 refresh에 관한 socket.on 제거

- 과거에 정의해놓고 사용하지 않고 있었음

voidmatcha committed:
fix: 🐛 소켓 시그널링 disconnect시 클라이언트에서 제대로 동작하도록 수정

- useEffect(() => {}, []); 에서 socket.on시 useState 상태를 못가져와서 useRef로 처리
- 예전에 작성해둔 length가 0이고 step이 0보다 크면 첫 화면으로 돌가도록 하는 코드 없앰
- peers 상태는 id에 대한 값을 추가하여 객체배열형태로 만들었고 이를 PeerVideo 컴포넌트에도 적용
- userID를 peers에 넣을 때 불일치가 생겨서 userId를 userID로 변경
- 문법 오류 해결

voidmatcha commented on webpack.config.js:
```diff
@@ -122,6 +122,14 @@ module.exports = (env) => {
       ],
 
     devServer: {
+      proxy: {
+        '/socket': {
+          target: 'http://localhost:8000',
+          changeOrigin: true,
+          ws: true,
+          logLevel: 'debug',
+        },
+      },
```
webpack의 해당 proxy 설정은 CRA의 http proxy middleware의 역할을 동일하게 수행합니다.

동건님이 작성해주신 부분으로 local에서 socket을 이용해 WebRTC 시그널링을 가능하게끔 해줍니다.

voidmatcha commented on __mock-server/server.js:
```diff
@@ -1,33 +1,42 @@
 const socket = require('socket.io');
 
-const io = socket(server);
+const io = socket(server, {
+  path: '/socket',
+  origins: ['http://localhost:3000', 'https://witherview-webrtc.herokuapp.com/'],
+
+  handlePreflightRequest: (req, res) => {
+    res.writeHead(200, {
+      'Access-Control-Allow-Origin': '*',
+      'Access-Control-Allow-Methods': 'GET,POST',
+      'Access-Control-Allow-Credentials': true,
+    });
+    res.end();
+  },
+});
```
이 부분도 마찬가지인데 이렇게 설정을 해주어야 local에서 socket을 이용해 WebRTC 시그널링을 가능하게끔 해줍니다.

voidmatcha commented on __mock-server/server.js:
```diff
@@ -49,28 +58,33 @@ io.on('connection', (sock) => {
+    // 서버에 저장된 데이터 삭제.
+    socketToRoom.delete(sock.id);
+
+    const idx = users[roomId].findIndex((item) => item === sock.id);
+
+    users[roomId].splice(idx, 1);
```
disconnect 된 경우 해당 sock.id에 대해서 users 객체에서도 데이터를 삭제해야하기 때문에 필요한 코드입니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -1,12 +1,22 @@
 export default function useSocketSignal({ roomId, setStep }) {
   const [peers, setPeers] = useState([]);
+
+  // useEffect로 마운트 실행시 socket.on에서 useState로 바뀐 상태를 가져오지 못하는 문제가 있어서 아래와 같이 처리
+  // TODO: 더 좋은 방법이 있다면 그것을 사용
+  const peersStateTempRef = useRef();
+  useEffect(() => {
+    peersStateTempRef.current = peers;
+  }, [peers]);
```
문제의 부분입니다. useEffect(fn, []) 로 마운트 시점에 socket.on을 해두고 서버로부터 emit 받으면 useState로 정의되어 변경된 상태를 그대로 가져오지 못하고 이전 시점의 상태값을 가져옵니다. 따라서 useRef와 useEffect를 이용해 peers의 상태값이 변경되면 peersStateTempRef에 담도록 하였습니다.

https://github.com/witherview/witherview_frontend/pull/3#discussion_r584706180

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -43,7 +53,10 @@ export default function useSocketSignal({ roomId, setStep }) {
 
   useEffect(() => {
-    socketRef.current = io.connect('https://witherview.herokuapp.com/');
+    socketRef.current = io.connect('/', {
+      transports: ['websocket'],
+      path: '/socket',
+    });
```
처음에 언급했던것과 마찬가지로 이렇게 설정을 해주어야 local에서 socket을 이용해 WebRTC 시그널링을 가능하게끔 해줍니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -54,40 +67,51 @@ export default function useSocketSignal({ roomId, setStep }) {
+          setPeers(peers);
+        });
+
+        socketRef.current.on('room full', () => {
+          history.push('/peer-study');
         });
```
서버에서 room full을 emit하면 클라이언트에서 그룹 스터디 초기화면으로 가도록 처리하였습니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -54,40 +67,51 @@ export default function useSocketSignal({ roomId, setStep }) {
 
         socketRef.current.on('all users', (users) => {
           setPeers([]);
-          const getPeers = [];
           users.forEach((userID) => {
             const peer = createPeer(userID, socketRef.current.id, stream);
-            peersRef.current.push({
-              peerID: userID,
-              peer,
-            });
-            getPeers.push(peer);
+            peersRef.current.set(userID, peer);
+            peers.push({ userID, peer });
```
userID와 peer를 포함한 객체를 객체배열 형태로 상태를 저장하도록 변경하였습니다. 이렇게 변경한 이유는           
```javascript
const remainingUsers = peersStateTempRef.current
            .filter((user) => {
              if (user.userID === leftSocketId) {
                return undefined;
              }
              return user;
            });
```
위와 같이 disconnect시 emit된 sock.id와 일치하는 peer id를 찾아야할 필요성이 있기 때문입니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -54,40 +67,51 @@ export default function useSocketSignal({ roomId, setStep }) {
-            setPeers((users) => [...users, peer]);
+            const userID = payload.callerID;
+            peersRef.current.set(userID, peer);
+            setPeers((users) => [...users, { userID, peer }]);
```
이부분도 같은 이유로 객체를 넣도록 하였습니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -54,40 +67,51 @@ export default function useSocketSignal({ roomId, setStep }) {
-          item.peer.signal(payload.signal);
+          const signalSenderId = payload.id;
+          const item = peersRef.current.get(signalSenderId);
+          if (item !== undefined) {
+            item.signal(payload.signal);
+          }
         });
+        // 누군가가 방을 떠났다는 메시지를 받았을 때
+        socketRef.current.on('user left', (leftSocketId) => {
+          // 떠난 유저를 제외한 나머지 배열을 반환한 뒤 setPeers로 다시 세팅
+          if (peersRef.current.get(leftSocketId) !== undefined) {
+            peersRef.current.delete(leftSocketId);
+          }
+          // 연결한 peer connection 삭제
+          const remainingUsers = peersStateTempRef.current
+            .filter((user) => {
+              if (user.userID === leftSocketId) {
+                return undefined;
+              }
+              return user;
+            });
 
-        socketRef.current.on('refresh', () => {
-          setPeers([]);
-          setStep(0);
+          setPeers(remainingUsers);
         });
```
disconnect 로직입니다. 1:1 상황에서는 완벽하게 동작하지만 1:N 상황에서 disconnect시 일부 유저들이 스트림을 가져오지 못하는 이슈가 발생합니다. console.log로 찍어보면 peers에 들어있는 접속된 유저 id가 알맞게 존재하므로 정상이기때문에 다른 부분에 문제가 있을 것으로 판단됩니다.

--- 

굳이 peersStateTempRef.current를 사용해 filter를 하는 이유는 앞서 언급했듯이 useState의 상태를 가져오지 못하는 문제가 있기 때문입니다. 

https://rinae.dev/posts/a-complete-guide-to-useeffect-ko 글에서 useEffect(fn, [])에 대해 `만약 “최신의” 상태 등을 원한다면, ref에다 담아둘 수는 있습니다.` 라는 문장을 읽고 useRef를 사용하여 처리하였습니다.

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerStudyTrainPage.js:
```diff
@@ -80,7 +80,7 @@ export default function PeerStudyTrainPage({
       setStep(STEP_CONNECT);
     }
-    if (peers.length === 0 && step > 0) history.push('/peer-study');
+    // if (peers.length === 0 && step > 0) history.push('/peer-study');
```
해당 코드는 필요 없어 주석처리하였습니다.

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerVideo.js:
```diff
@@ -34,7 +34,7 @@ export default function PeerVideo({
 
   useEffect(() => {
-    peer.on('stream', (stream) => {
+    peer.peer.on('stream', (stream) => {
```
앞서 peers 상태는 userID를 포함하는 객체배열 형태로 수정하였기 때문에 추가된 부분입니다.

voidmatcha reviewed (commented):
1:1 상황에서는 join 부터 disconnect까지 로컬에서 완벽하게 동작합니다. 다만 1:N 상황에서 정상적으로 동작해야할 상황임에도 스트림을 가져오지 못하는 이슈가 있어 이를 더 확인해봐야 할 것 같습니다.

voidmatcha committed:
refactor: 💡 signaling server RTCPeerConnection 에 맞게 변경한다 (#3)

- on, emit 하는 부분 RTCPeerConnection에 맞도록 수정
  - offer, answer, candidate

voidmatcha committed:
refactor: 💡 배포된 url로 webpack proxy target url을 변경한다 (#3)

- 로컬 url에서 heroku로 배포한 url로 변경함

voidmatcha committed:
refactor: 💡 simple peer에서 RTCPeerConnection로 구조 변경한다 (#3)

- 추상화된 라이브러리인 simple peer를 사용했었는데 이를 걷어냄
- RTCPeerConnection를 사용해서 WebRTC 통신이 가능하도록 함
  - 이에 맞게 전체적인 구조 변경
- 각 peer의 계정 상태를 들고 있게끔 구조 변경

voidmatcha commented on __mock-server/server.js:
```diff
@@ -8,7 +7,7 @@ const socket = require('socket.io');
 const io = socket(server, {
   path: '/socket',
-  origins: ['http://localhost:3000', 'https://witherview-webrtc.herokuapp.com/'],
+  origins: ['http://localhost:3000', 'https://witherview-test.herokuapp.com/'],
```
@inspirit941 배포 후에도 정상 작동하는 것을 확인했습니다. 마침 AWS Active에 선정되었으니 AWS 인프라로 배포해주시면 해당 URL로 변경하도록 하겠습니다.

voidmatcha commented on __mock-server/server.js:
```diff
@@ -25,66 +24,84 @@ const users = {};
+    console.log(usersInThisRoom);
+
+    io.sockets.to(sock.id).emit('all users', usersInThisRoom);
   });
```
users 객체에 id와 유저 계정 정보를 포함할 수 있도록 하였습니다.

voidmatcha commented on __mock-server/server.js:
```diff
@@ -25,66 +24,84 @@ const users = {};
+      candidate,
+      candidateSendID,
+    });
   });
```
추상화한 라이브러리인 simple peer를 걷어내고 순수하게 Web API(RTCPeerConnection 등)를 이용해 시그널링을 하는 상황에 맞게 수정하였습니다. 

각각은 아래의 이미지의 절차를 수행하게 됩니다.

![image](https://user-images.githubusercontent.com/16266103/113164006-ac0b8180-927b-11eb-817c-d6f9777c412c.png)

voidmatcha commented on __mock-server/server.js:
```diff
@@ -25,66 +24,84 @@ const users = {};
+        delete users[roomId];
+      }
+    }
+  });
```
disconnect는 이제 3명 이상이 접속한 경우에도 문제없이 잘 동작 합니다.

voidmatcha commented on webpack.config.js:
```diff
@@ -124,7 +124,7 @@ module.exports = (env) => {
       proxy: {
         '/socket': {
-          target: 'http://localhost:8000',
+          target: 'https://witherview-test.herokuapp.com',
```
이 부분도 마찬가지로 AWS 인프라로 배포해주시면 해당 URL로 변경하도록 하겠습니다.

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerVideo.js:
```diff
@@ -34,10 +34,8 @@ export default function PeerVideo({
-    });
-  }, []);
+    ref.current.srcObject = peer.stream;
+  }, [peer]);
```
이부분이 문제의 근원이라고 봐도 되는데 simple peer를 WebAPI로 변경하면서 발견하였습니다.

3명 이상 접속했을 경우 누군가 나갔을 때 상대 peer의 stream이 검은화면으로 바뀌는 이슈를 해결하기 위한 부분입니다.

useSocketSignal의 peers의 상태가 바뀐 경우를 고려하기 위해 빈 배열이 아닌 peer가 바뀌는 상황에서 재실행되게끔 했어야 이슈 없이 의도대로 동작하게 됩니다.

사실 이를 적용하면 simple peer에서 해결하지 못했던 disconnect 이슈를 해결할 수 있습니다. 다만, WebAPI로 바꿔놓은 형태가 WebRTC의 구조를 바탕으로 파악하는데 더 적합하다고 생각하여 simple peer를 더이상 사용하지 않는 방향으로 결론을 내렸습니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -1,131 +1,217 @@
-          history.push('/peer-study');
-        });
+  const getStream = async () => {
+    try {
```
기존 코드에서 promise로 되어있는 부분을 async await으로 변경하였습니다. 에러 핸들링은 try catch로 처리하였습니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -1,131 +1,217 @@
-              }
-              return user;
-            });
+      socketRef.current.emit('join room', {
+        roomId,
+        email,
+      });
```
로그인된 유저의 계정을 포함해서 시그널링을 하게끔 처리해두었습니다.

voidmatcha reviewed (commented):
기존에 있었던 문제들을 해결하였습니다.

1. 서버 배포시 에러가 발생했던 부분이 해결되었습니다. (현재 heroku를 통해 배포되어 있습니다)
2. 3명 이상 peer가 연결되었을 때 누군가 나갈경우 불특정 peer의 stream의 검은 화면으로 바뀌는 이슈를 해결되었습니다.

리뷰 및 테스트 후 approve 해주시면 감사하겠습니다

voidmatcha committed:
Merge branch 'develop' of https://github.com/witherview/witherview_frontend into feature/backend
